tenable/ad/roles/api.py

```diff
@@ -0,0 +1,235 @@
+            ...     )
+        '''
+        self._delete(f"{role_id}")
+
+    def copy_role(self,
+                  from_id: str,
+                  name: str
+                  ) -> Dict:
+        '''
+        Creates a new role from another role
+
+        Args:
+            from_id (str):
+                The role instance identifier user wants to copy.
+            name (str):
+                The name of new role.
+
+        Returns:
+            dict:
+                the copied role object.
+
+        Examples:
+            >>> tad.roles.copy_role(
+            ...     from_id='1',
+            ...     name='Copied name'
+            ...     )
+        '''
+        payload = self._schema.dump(self._schema.load({
+            'name': name
+        }))
+        return self._schema.load(self._post(f'from/{from_id}', json=payload))
+
+    def replace_role_permissions(self,
+                                 role_id: str,
+                                 permissions: List[Dict]
+                                 ) -> Dict:
+        '''
+        Replace permission list for a role
+
+        Args:
+            role_id (str):
+                The role instance identifier.
+            permissions (List[Dict]) :
+                The list of permissions dictionaries.
+                Below are the values expected in dictionaries
+            entity_name (str):
+                The name of entity.
+            action (str):
+                ???
+            entity_ids (List[int]):
+                The list of entity identifiers.
+            dynamic_id (optional, str):
+                ???
+
+        Returns:
+            dict:
+                the update permissions role object.
+
+        Examples:
+            >>> tad.roles.replace_role_permissions(
+            ...     role_id='1',
+            ...     permissions=[{
+            ...         'entity_name':'dashboard',
+            ...         'action':'action',
+            ...         'entity_ids':[1, 2],
+            ...         'dynamic_id': None
+            ...     }]
+            ... )
+        '''
+        schema = RolePermissionsSchema()
+
+        payload = []
+        required_fields = ['action', 'entity_ids', 'entity_name']
+
+        # loop through all permissions and create payload for API call
+        # and if any of required fields is missing then it will raise
+        # validationError
+        for permission in permissions:
+            if all(key in permission for key in required_fields):
+                payload.append({
+                    'dynamicId': permission.get('dynamic_id'),
+                    'entityIds': permission.get('entity_ids'),
+                    'entityName': permission.get('entity_name'),
+                    'action': permission.get('action')
+                })
+            else:
+                raise ValidationError(
+                    'Missing required fields. '
+                    'Required: ["action", "entity_name", "entity_ids"]')
```

Comment:
Why not rely on marshmallow to handle the field requirements?

Reply:
updated code to handle required fields on marshmallow schema